authentication/lib/src/login_page.dart

```diff
@@ -0,0 +1,207 @@
+import 'package:firebase_auth/firebase_auth.dart';
+import 'package:flutter/material.dart';
+
+class LoginPage extends StatefulWidget {
+  const LoginPage({Key? key}) : super(key: key);
+
+  @override
+  State<LoginPage> createState() => _LoginPageState();
+}
+
+class _LoginPageState extends State<LoginPage> {
+  final TextEditingController _usernameController = TextEditingController();
+  final TextEditingController _passwordController = TextEditingController();
+  final TextEditingController _firstnameController = TextEditingController();
+  final TextEditingController _lastnameController = TextEditingController();
+  bool _isLoading = false;
+  bool _isSigningUp = false;
+  final _formKey = GlobalKey<FormState>();
+
+  void _setIsLoading() {
+    setState(() {
+      _isLoading = !_isLoading;
+    });
+  }
+
+  Future<UserCredential?> _signUpWithEmailAndPassword() async {
+    _setIsLoading();
+    try {
+      final _credential =
+          await FirebaseAuth.instance.createUserWithEmailAndPassword(
+        email: _usernameController.text,
+        password: _passwordController.text,
+      );
+
+      await _credential.user!.updateDisplayName(
+          '${_firstnameController.text} ${_lastnameController.text}');
+
+      return _credential;
+    } on FirebaseAuthException catch (e) {
+      if (e.code == 'invalid-email') {
+        // ... ask user to update email value
+      }
+      if (e.code == 'weak-password') {
+        // ...
+      }
+      _setIsLoading();
+      return null;
+    }
+  }
+
+  Future<UserCredential?> _loginWithEmailAndPassword() async {
+    _setIsLoading();
+
+    try {
+      return await FirebaseAuth.instance.signInWithEmailAndPassword(
+        email: _usernameController.text,
+        password: _passwordController.text,
+      );
+    } on FirebaseAuthException catch (e) {
+      if (e.code == 'user-not-found') {
+        // ...
+        showDialog(
+            context: context,
+            builder: (context) {
+              return AlertDialog(
+                title: const Text('Email doesn\'t exist'),
+                content: const Text('Please sign up instead'),
+                actions: [
+                  ElevatedButton(
+                    onPressed: () => Navigator.of(context).pop(),
+                    child: const Text('OK'),
+                  ),
+                ],
+              );
+            });
+      }
+      if (e.code == 'email-doesnt-exist') {
```

Comment:
I love that you're hitting so many of these codes. Still, even if you're actually covering them all here, for peace of mind of the reader, I think adding a link to whatever the best documentation on these is (could be the raw code in the SDK) would be a nice touch.